masonry_core/src/passes/update.rs

```diff
@@ -451,56 +442,119 @@ fn update_focus_chain_for_widget(
     let widget = &mut **widget.item;
     let state = state.item;
 
-    // Replace has_focused to check if the value changed in the meantime
-    state.has_focus_target = global_state.focused_widget == Some(id);
-    let had_focus = state.has_focus_target;
-
-    if state.needs_update_focus_chain {
-        state.focus_chain.clear();
-        if state.accepts_focus {
-            state.focus_chain.push(id);
-        }
-        state.needs_update_focus_chain = false;
-
-        let parent_state = &mut *state;
-        recurse_on_children(id, widget, children, |widget, mut state, properties| {
-            update_focus_chain_for_widget(
-                global_state,
-                widget,
-                state.reborrow_mut(),
-                properties,
-                &mut parent_state.focus_chain,
-            );
-            parent_state.merge_up(state.item);
-        });
+    if !state.needs_update_focusable {
+        return;
     }
 
-    if !state.is_disabled && !state.is_stashed {
-        parent_focus_chain.extend(&state.focus_chain);
-    }
+    state.descendant_is_focusable = false;
 
-    // had_focus is the old focus value. state.has_focused was replaced with parent_ctx.is_focused().
-    // Therefore if had_focus is true but state.has_focused is false then the widget which is
-    // currently focused is not part of the functional tree anymore and should resign the focus.
-    if had_focus && !state.has_focus_target {
-        // Not sure about this logic, might remove
-        global_state.next_focused_widget = None;
+    if state.accepts_focus {
+        state.descendant_is_focusable = true;
     }
-    state.has_focus_target = had_focus;
+
+    let parent_state = &mut *state;
+    recurse_on_children(id, widget, children, |widget, mut state, properties| {
+        update_focusable_for_widget(widget, state.reborrow_mut(), properties);
+
+        if state.item.descendant_is_focusable {
+            parent_state.descendant_is_focusable = true;
+        }
+    });
+
+    state.needs_update_focusable = false;
 }
 
 pub(crate) fn run_update_focus_chain_pass(root: &mut RenderRoot) {
     let _span = info_span!("update_focus_chain").entered();
-    let mut dummy_focus_chain = Vec::new();
 
     let (root_widget, root_state, root_properties) = root.widget_arena.get_all_mut(root.root.id());
-    update_focus_chain_for_widget(
-        &mut root.global_state,
-        root_widget,
-        root_state,
-        root_properties,
-        &mut dummy_focus_chain,
-    );
+    update_focusable_for_widget(root_widget, root_state, root_properties);
+}
+
+pub(crate) fn find_next_in_focus_chain(root: &mut RenderRoot, forward: bool) -> Option<WidgetId> {
+    let focus_anchor = root.global_state.focus_anchor;
+
+    // If the anchor isn't already focused, try to find its first focusable descendant.
+    if focus_anchor != root.global_state.focused_widget
```

Comment:
What about if we are focused on the focus anchor (and are going forwards)? For a tree like:
```text
A
  - B [focused]
    - C
  - D
E
```
Where all five nodes are focusable, and the focus is currently on B, I would expect the focus order to be:
`B -> C -> D -> E`.

(This also needs care when going backwards)